parquet-column/src/main/java/org/apache/parquet/column/values/bloom/Bloom.java

```diff
@@ -0,0 +1,484 @@
+      mask[i] = mask[i] >> 27;
+    }
+
+    for (int i = 0; i < 8; ++i) {
+      mask[i] = 0x1 << mask[i];
+    }
+
+  }
+
+  /**
+   * Add an element to bloom filter, the element content is represented by
+   * the hash value of its plain encoding result.
+   * @param hash hash result of element.
+   */
+  public void bloomInsert(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (numBytes / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      int bitsetIndex = bucketIndex * BYTES_PER_BUCKET + i * 4;
+      bitset[bitsetIndex] |= (byte)(mask[i] >>> 24);
+      bitset[bitsetIndex + 1] |= (byte)(mask[i] >>> 16);
+      bitset[bitsetIndex + 2] |= (byte)(mask[i] >>> 8);
+      bitset[bitsetIndex + 3] |= (byte)(mask[i]);
+    }
+  }
+
+  /**
+   * Determine where an element is in set or not.
+   * @param hash the hash value of element plain encoding result.
+   * @return false if element is must not in set, true if element probably in set.
+   */
+  public boolean bloomFind(long hash) {
+    int bucketIndex = (int)(hash >> 32) & (numBytes / BYTES_PER_BUCKET - 1);
+    int key = (int)hash;
+
+    setMask(key);
+
+    for (int i = 0; i < 8; i++) {
+      byte set = 0;
+      int bitsetIndex = bucketIndex * BYTES_PER_BUCKET + i * 4;
+      set |= bitset[bitsetIndex] & ((byte)(mask[i] >>> 24));
+      set |= bitset[bitsetIndex + 1] & ((byte)(mask[i] >>> 16));
+      set |= bitset[bitsetIndex + 2] & ((byte)(mask[i] >>> 8));
+      set |= bitset[bitsetIndex + 3] & ((byte)mask[i]);
+      if (0 == set) {
+        return false;
+      }
+    }
+
+    return true;
+  }
+
+  /**
+   * Bloom filter bitset can be created lazily, flush() will set bits for
+   * all elements in cache. If bitset was already created and set, it do nothing.
+   */
+  public void flush() {
+    if (!elements.isEmpty() && bitset == null) {
+      initBitset(optimalNumOfBits(elements.size(), FPP)/8);
+      for (long hash : elements) {
+        bloomInsert(hash);
+      }
+      elements.clear();
+    }
+  }
+
+  /**
+   * Calculate optimal size according to the number of distinct values and false positive probability.
+   * @param n: The number of distinct values.
+   * @param p: The false positive probability.
+   * @return optimal number of bits of given n and p.
+   */
+  public static int optimalNumOfBits(long n, double p) {
+    int bits = (int)(-n * Math.log(p) / (Math.log(2) * Math.log(2)));
+
+    bits --;
+    bits |= bits >> 1;
+    bits |= bits >> 2;
+    bits |= bits >> 4;
+    bits |= bits >> 8;
+    bits |= bits >> 16;
+    bits++;
+
+    return bits;
+  }
+
+
+  /**
+   * Element is represented by hash in bloom filter. The hash function takes plain encoding
+   * of element as input.
+   */
+  public Encoding getEncoding() {
+    return Encoding.PLAIN;
+  }
+
+  /**
+   * @return Bytes buffered of bloom filter.
+   */
+  public long getBufferedSize() {
+    return elements.size() * 8 + numBytes;
```

Comment:
1. What do you mean by buffered bytes?

2. Also, why multiply by 8?

Reply:
1. Buffered bytes is the total number of bytes we allocate in heap for bloom filter, please refer to similar definition in [dictionaryValueWriter](https://github.com/apache/parquet-mr/blob/master/parquet-column/src/main/java/org/apache/parquet/column/values/dictionary/DictionaryValuesWriter.java#L149).

2. Elements are stored in Long which is 8 bytes.

Reply:
1. Please add a comment that explains it as well as the method on https://github.com/apache/parquet-mr/blob/70f28810a5547219e18ffc3465f519c454fee6e5/parquet-column/src/main/java/org/apache/parquet/column/values/ValuesWriter.java

2. A HashSet<Long> will use a good bit more than 8 heap bytes per value.

Reply:
What do you mean of "A HashSet will use a good bit more than 8 heap bytes per value."?

Reply:
I'm not sure how to say it another way. A HashSet<Long> containing N Longs will use substantially more than N*8 bytes of memory. This is because of object overhead, empty space, and other overheads associated with hash tables occupy heap space.

Reply:
understood!